server/memphis_handlers_tags.go

```diff
@@ -0,0 +1,467 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+package server
+
+import (
+	"context"
+	"errors"
+	"memphis-broker/models"
+	"memphis-broker/utils"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+type TagsHandler struct{ S *Server }
+
+const (
+	tagObjectName = "Tag"
+)
+
+func CreateTag(name string, from string, from_name string, background_color string, text_color string) error {
+	name = strings.ToLower(name)
+	exist, _, err := IsTagExist(name)
+	if err != nil {
+		return err
+	}
+	if exist {
+		return errors.New("Tag with that name already exists")
+	}
+	var newTag models.Tag
+	stationArr := []primitive.ObjectID{}
+	schemaArr := []primitive.ObjectID{}
+	userArr := []primitive.ObjectID{}
+	switch from {
+	case "station":
+		station_name, err := StationNameFromStr(from_name)
+		if err != nil {
+			return err
+		}
+		exist, station, err := IsStationExist(station_name)
+		if err != nil {
+			return err
+		}
+		if !exist {
+			return errors.New("Station with this name does not exist")
+		}
+		stationArr = append(stationArr, station.ID)
+		// case "schema":
+		// 	exist, schema, err := IsSchemaExist(from_name)
+		// 	if err != nil {
+		// 		return err
+		// 	}
+		// 	if !exist {
+		// 		return errors.New("Schema with this name does not exist")
+		// 	}
+		// 	schemaArr = append(schemaArr, schema.ID)
+		// case "user":
+		// 	exist, user, err := IsUserExist(from_name)
+		// 	if err != nil {
+		// 		return err
+		// 	}
+		// 	if !exist {
+		// 		return errors.New("User with this name does not exist")
+		// 	}
+		// 	userArr = append(userArr, user.ID)
+	}
+	newTag = models.Tag{
+		ID:       primitive.NewObjectID(),
+		Name:     name,
+		ColorBG:  background_color,
+		ColorTXT: text_color,
+		Stations: stationArr,
+		Schemas:  schemaArr,
+		Users:    userArr,
+	}
+	_, err = tagsCollection.InsertOne(context.TODO(), newTag)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func AddTagToEntity(name string, to string, to_name string, background_color string, text_color string) error {
+	name = strings.ToLower(name)
+	exist, tag, err := IsTagExist(name)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		err = CreateTag(name, to, to_name, background_color, text_color)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	switch to {
+	case "station":
+		station_name, err := StationNameFromStr(to_name)
+		if err != nil {
+			return err
+		}
+		exist, station, err := IsStationExist(station_name)
+		if err != nil {
+			return err
+		}
+		if !exist {
+			return errors.New("Station with this name does not exist")
+		}
+		_, err = tagsCollection.UpdateOne(context.TODO(), bson.M{"_id": tag.ID}, bson.M{"$addToSet": bson.M{"stations": station.ID}})
```

Comment:
what are the differences between $addToSet to $push, why not use $push

Reply:
If the value already exists in the array, $push will still append the value (resulting in duplicate values). However, $addToSet only appends the value if it doesn't already exist in the array